File: cmd/llmash/progress.go
package main

import (
	"bufio"
	"fmt"
	"io"
	"os"
	"strings"
	"sync"
	"sync/atomic"
	"time"
)

// Ollama's progress display: a set of lines (spinners and bars) on stderr,
// redrawn in place ten times a second.

const (
	defaultTermWidth  = 80
	defaultTermHeight = 24
)

type progState interface{ String() string }

type progress struct {
	mu       sync.Mutex
	w        *bufio.Writer
	pos      int
	states   []progState
	stopOnce sync.Once
	done     chan struct{}
}

func newProgress(w io.Writer) *progress {
	p := &progress{w: bufio.NewWriter(w), done: make(chan struct{})}
	go p.run()
	return p
}

func (p *progress) add(state progState) {
	p.mu.Lock()
	defer p.mu.Unlock()
	p.states = append(p.states, state)
}

func (p *progress) halt() (bool, int) {
	var stopped bool
	p.stopOnce.Do(func() {
		close(p.done)
		stopped = true
	})
	p.mu.Lock()
	defer p.mu.Unlock()
	for _, s := range p.states {
		if sp, ok := s.(*progSpinner); ok {
			sp.stop()
		}
	}
	if stopped {
		p.renderLocked()
	}
	return stopped, p.pos
}

func (p *progress) stop() bool {
	stopped, _ := p.halt()
	if stopped {
		fmt.Fprint(p.w, "\n")
		p.w.Flush()
	}
	return stopped
}

func (p *progress) stopAndClear() bool {
	defer p.w.Flush()
	fmt.Fprint(p.w, "\033[?25l")
	defer fmt.Fprint(p.w, "\033[?25h")
	stopped, pos := p.halt()
	if stopped {
		for i := 0; i < pos; i++ {
			if i > 0 {
				fmt.Fprint(p.w, "\033[A")
			}
			fmt.Fprint(p.w, "\033[2K\033[1G")
		}
	}
	return stopped
}

func (p *progress) render() {
	p.mu.Lock()
	defer p.mu.Unlock()
	p.renderLocked()
}

func (p *progress) renderLocked() {
	defer p.w.Flush()
	fmt.Fprint(p.w, "\033[?2026h")
	defer fmt.Fprint(p.w, "\033[?2026l")
	fmt.Fprint(p.w, "\033[?25l")
	defer fmt.Fprint(p.w, "\033[?25h")
	for i := 0; i < p.pos-1; i++ {
		fmt.Fprint(p.w, "\033[A")
	}
	fmt.Fprint(p.w, "\033[1G")
	height := defaultTermHeight
	max := len(p.states)
	if max > height {
		max = height
	}
	for i := len(p.states) - max; i < len(p.states); i++ {
		fmt.Fprint(p.w, p.states[i].String(), "\033[K")
		if i < len(p.states)-1 {
			fmt.Fprint(p.w, "\n")
		}
	}
	p.pos = len(p.states)
}

func (p *progress) run() {
	t := time.NewTicker(100 * time.Millisecond)
	defer t.Stop()
	for {
		select {
		case <-p.done:
			return
		case <-t.C:
			p.render()
		}
	}
}

// ------------------------------------------------------------- spinner

type progSpinner struct {
	message      atomic.Value
	mu           sync.Mutex
	messageWidth int
	parts        []string
	value        int
	stopped      time.Time
	stopOnce     sync.Once
	done         chan struct{}
}

func newSpinner(message string) *progSpinner {
	s := &progSpinner{
		parts: []string{"⠋", "⠙", "⠹", "⠸", "⠼", "⠴", "⠦", "⠧", "⠇", "⠏"},
		done:  make(chan struct{}),
	}
	s.message.Store(message)
	go s.run()
	return s
}

func (s *progSpinner) String() string {
	s.mu.Lock()
	defer s.mu.Unlock()
	var sb strings.Builder
	if m, ok := s.message.Load().(string); ok && len(m) > 0 {
		m = strings.TrimSpace(m)
		if s.messageWidth > 0 && len(m) > s.messageWidth {
			m = m[:s.messageWidth]
		}
		sb.WriteString(m)
		if pad := s.messageWidth - sb.Len(); pad > 0 {
			sb.WriteString(strings.Repeat(" ", pad))
		}
		sb.WriteString(" ")
	}
	if s.stopped.IsZero() {
		sb.WriteString(s.parts[s.value])
		sb.WriteString(" ")
	}
	return sb.String()
}

func (s *progSpinner) run() {
	t := time.NewTicker(100 * time.Millisecond)
	defer t.Stop()
	for {
		select {
		case <-s.done:
			return
		case <-t.C:
			s.mu.Lock()
			s.value = (s.value + 1) % len(s.parts)
			s.mu.Unlock()
		}
	}
}

func (s *progSpinner) stop() {
	s.mu.Lock()
	if s.stopped.IsZero() {
		s.stopped = time.Now()
	}
	s.mu.Unlock()
	s.stopOnce.Do(func() { close(s.done) })
}

// ----------------------------------------------------------------- bar

type progBar struct {
	mu           sync.Mutex
	message      string
	messageWidth int
	maxValue     int64
	initialValue int64
	currentValue int64
	started      time.Time
	stopped      time.Time
	maxBuckets   int
	buckets      []progBucket
}

type progBucket struct {
	updated time.Time
	value   int64
}

func newBar(message string, maxValue, initialValue int64) *progBar {
	b := &progBar{message: message, messageWidth: -1, maxValue: maxValue,
		initialValue: initialValue, currentValue: initialValue, started: time.Now(), maxBuckets: 10}
	if initialValue >= maxValue {
		b.stopped = time.Now()
	}
	return b
}

// Two units at most: "1h20m", "45s".
func formatDuration(d time.Duration) string {
	switch {
	case d >= 100*time.Hour:
		return "99h+"
	case d >= time.Hour:
		return fmt.Sprintf("%dh%dm", int(d.Hours()), int(d.Minutes())%60)
	default:
		return d.Round(time.Second).String()
	}
}

func (b *progBar) String() string {
	w := termWidthOf(os.Stderr)
	if w <= 0 {
		w = defaultTermWidth
	}
	b.mu.Lock()
	defer b.mu.Unlock()

	var pre strings.Builder
	if len(b.message) > 0 {
		m := strings.TrimSpace(b.message)
		if b.messageWidth > 0 && len(m) > b.messageWidth {
			m = m[:b.messageWidth]
		}
		pre.WriteString(m)
		if pad := b.messageWidth - pre.Len(); pad > 0 {
			pre.WriteString(strings.Repeat(" ", pad))
		}
		pre.WriteString(" ")
	}
	fmt.Fprintf(&pre, "%3.0f%%", b.percent())

	var suf strings.Builder
	pad := func(s string, n int) {
		if k := n - len(s); k > 0 {
			suf.WriteString(strings.Repeat(" ", k))
		}
		suf.WriteString(s)
	}
	if b.stopped.IsZero() {
		pad(humanBytes(b.currentValue), 6)
		suf.WriteString("/")
		pad(humanBytes(b.maxValue), 6)
	} else {
		pad(humanBytes(b.maxValue), 6)
		suf.WriteString(strings.Repeat(" ", 7))
	}
	rate := b.rate()
	if b.stopped.IsZero() && rate > 0 {
		suf.WriteString("  ")
		pad(humanBytes(int64(rate)), 6)
		suf.WriteString("/s")
	} else {
		suf.WriteString(strings.Repeat(" ", 10))
	}
	if b.stopped.IsZero() && rate > 0 {
		suf.WriteString("  ")
		remaining := time.Duration(int64(float64(b.maxValue-b.currentValue)/rate)) * time.Second
		pad(formatDuration(remaining), 6)
	} else {
		suf.WriteString(strings.Repeat(" ", 8))
	}

	var mid strings.Builder
	f := w - pre.Len() - suf.Len() - 5
	if f < 0 {
		f = 0
	}
	n := int(float64(f) * b.percent() / 100)
	mid.WriteString(" ▕")
	if n > 0 {
		mid.WriteString(strings.Repeat("█", n))
	}
	if f-n > 0 {
		mid.WriteString(strings.Repeat(" ", f-n))
	}
	mid.WriteString("▏ ")
	return pre.String() + mid.String() + suf.String()
}

func (b *progBar) set(value int64) {
	b.mu.Lock()
	defer b.mu.Unlock()
	if value >= b.maxValue {
		value = b.maxValue
	}
	b.currentValue = value
	if b.currentValue >= b.maxValue {
		b.stopped = time.Now()
	}
	if len(b.buckets) == 0 || time.Since(b.buckets[len(b.buckets)-1].updated) > time.Second {
		b.buckets = append(b.buckets, progBucket{updated: time.Now(), value: value})
		if len(b.buckets) > b.maxBuckets {
			b.buckets = b.buckets[1:]
		}
	}
}

func (b *progBar) percent() float64 {
	if b.maxValue > 0 {
		return float64(b.currentValue) / float64(b.maxValue) * 100
	}
	return 0
}

func (b *progBar) rate() float64 {
	var num, den float64
	if !b.stopped.IsZero() {
		num = float64(b.currentValue - b.initialValue)
		den = b.stopped.Sub(b.started).Round(time.Second).Seconds()
	} else if len(b.buckets) > 1 {
		first, last := b.buckets[0], b.buckets[len(b.buckets)-1]
		num = float64(last.value - first.value)
		den = last.updated.Sub(first.updated).Round(time.Second).Seconds()
	}
	if den == 0 {
		return 0
	}
	return num / den
}

File: cpp/src/progress.cpp
#include "progress.h"

#include "cli_format.h"
#include "terminal.h"

#include <algorithm>
#include <cmath>
#include <cstdio>
#include <cstring>

namespace llmash {
namespace {

constexpr int kTermWidth  = 80;
constexpr int kTermHeight = 24;

const char * const kSpinnerParts[] = {"⠋", "⠙", "⠹", "⠸", "⠼",
                                      "⠴", "⠦", "⠧", "⠇", "⠏"};
constexpr size_t kSpinnerCount = sizeof(kSpinnerParts) / sizeof(kSpinnerParts[0]);

void out(const char * s) { std::fputs(s, stderr); }

std::string pad_left(const std::string & s, int n) {
    const int k = n - static_cast<int>(s.size());
    return k > 0 ? std::string(static_cast<size_t>(k), ' ') + s : s;
}

std::string repeat(const char * unit, int n) {
    std::string s;
    for (int i = 0; i < n; i++) {
        s += unit;
    }
    return s;
}

// Two units at most: "1h20m", "45s".
std::string format_duration(double seconds) {
    if (seconds >= 100 * 3600) {
        return "99h+";
    }
    const long long s = static_cast<long long>(std::llround(seconds));
    if (s >= 3600) {
        return std::to_string(s / 3600) + "h" + std::to_string((s % 3600) / 60) + "m";
    }
    if (s >= 60) {
        return std::to_string(s / 60) + "m" + std::to_string(s % 60) + "s";
    }
    return std::to_string(s) + "s";
}

double secs_since(ProgClock::time_point t) {
    return std::chrono::duration<double>(ProgClock::now() - t).count();
}

} // namespace

// ------------------------------------------------------------------ spinner

ProgSpinner::ProgSpinner(std::string message)
    : message_(std::move(message)), last_(ProgClock::now()) {}

void ProgSpinner::set_message(std::string m) {
    std::lock_guard<std::mutex> lock(mu_);
    message_ = std::move(m);
}

void ProgSpinner::stop() {
    std::lock_guard<std::mutex> lock(mu_);
    stopped_ = true;
}

std::string ProgSpinner::str() {
    std::lock_guard<std::mutex> lock(mu_);
    if (!stopped_ && secs_since(last_) >= 0.1) {
        value_ = (value_ + 1) % kSpinnerCount;
        last_  = ProgClock::now();
    }
    std::string s = message_;
    if (!s.empty()) {
        s += " ";
    }
    if (!stopped_) {
        s += kSpinnerParts[value_];
        s += " ";
    }
    return s;
}

// ---------------------------------------------------------------------- bar

ProgBar::ProgBar(std::string message, int64_t max_value, int64_t initial, ProgUnit unit)
    : message_(std::move(message)), max_value_(max_value), initial_(initial),
      current_(initial), unit_(unit), started_(ProgClock::now()) {
    if (initial >= max_value && max_value > 0) {
        stopped_    = true;
        stopped_at_ = ProgClock::now();
    }
}

double ProgBar::percent_locked() const {
    return max_value_ > 0 ? static_cast<double>(current_) / static_cast<double>(max_value_) * 100.0 : 0.0;
}

double ProgBar::rate_locked() const {
    double num = 0, den = 0;
    if (stopped_) {
        num = static_cast<double>(current_ - initial_);
        den = std::chrono::duration<double>(stopped_at_ - started_).count();
    } else if (buckets_.size() > 1) {
        const Bucket & first = buckets_.front();
        const Bucket & last  = buckets_.back();
        num = static_cast<double>(last.value - first.value);
        den = std::chrono::duration<double>(last.updated - first.updated).count();
    }
    return den > 0 ? num / den : 0.0;
}

void ProgBar::set(int64_t value) {
    std::lock_guard<std::mutex> lock(mu_);
    if (max_value_ > 0 && value >= max_value_) {
        value = max_value_;
    }
    current_ = value;
    if (max_value_ > 0 && current_ >= max_value_ && !stopped_) {
        stopped_    = true;
        stopped_at_ = ProgClock::now();
    }
    if (buckets_.empty() || secs_since(buckets_.back().updated) > 1.0) {
        buckets_.push_back({ProgClock::now(), value});
        if (buckets_.size() > 10) {
            buckets_.erase(buckets_.begin());
        }
    }
}

std::string ProgBar::str() {
    int w = terminal_columns();
    if (w <= 0) {
        w = kTermWidth;
    }
    std::lock_guard<std::mutex> lock(mu_);

    std::string pre = message_;
    if (!pre.empty()) {
        pre += " ";
    }
    char pct[16];
    std::snprintf(pct, sizeof(pct), "%3.0f%%", percent_locked());
    pre += pct;

    const auto figure = [&](int64_t v) {
        return unit_ == ProgUnit::Bytes ? human_bytes(v) : std::to_string(v);
    };
    std::string suf;
    if (!stopped_) {
        suf += pad_left(figure(current_), 6) + "/" + pad_left(figure(max_value_), 6);
    } else {
        suf += pad_left(figure(max_value_), 6) + std::string(7, ' ');
    }
    const double rate = rate_locked();
    if (!stopped_ && rate > 0) {
        // A count of tensors per second tells no one anything; the time left does.
        suf += unit_ == ProgUnit::Bytes ? "  " + pad_left(human_bytes(static_cast<int64_t>(rate)), 6) + "/s"
                                        : std::string(9, ' ');
        suf += "  " + pad_left(format_duration(static_cast<double>(max_value_ - current_) / rate), 6);
    } else {
        suf += std::string(18, ' ');
    }

    // The bar fills whatever the message and the figures leave, minus the
    // four cells the brackets and their spaces take.
    int f = w - static_cast<int>(pre.size()) - static_cast<int>(suf.size()) - 5;
    if (f < 0) {
        f = 0;
    }
    const int n = static_cast<int>(static_cast<double>(f) * percent_locked() / 100.0);
    std::string mid = " ▕";
    mid += repeat("█", n);
    mid += std::string(static_cast<size_t>(f - n), ' ');
    mid += "▏ ";
    return pre + mid + suf;
}

// ----------------------------------------------------------------- progress

Progress::Progress() {
    thread_ = std::thread([this] { run(); });
}

Progress::~Progress() {
    done_ = true;
    if (thread_.joinable()) {
        thread_.join();
    }
}

void Progress::add(const std::shared_ptr<ProgState> & state) {
    std::lock_guard<std::mutex> lock(mu_);
    states_.push_back(state);
}

void Progress::run() {
    while (!done_) {
        std::this_thread::sleep_for(std::chrono::milliseconds(100));
        if (!done_) {
            render();
        }
    }
}

void Progress::render() {
    std::lock_guard<std::mutex> lock(mu_);
    render_locked();
}

void Progress::render_locked() {
    if (!stderr_is_terminal()) {
        return;
    }
    // 2026 is synchronized output: the terminal shows the whole frame at once
    // instead of each line as it arrives, and 25l hides the caret, which is
    // what was flickering through every redraw.
    out("\x1b[?2026h");
    out("\x1b[?25l");
    for (int i = 0; i < pos_ - 1; i++) {
        out("\x1b[A");
    }
    out("\x1b[1G");
    const int max = std::min(static_cast<int>(states_.size()), kTermHeight);
    for (size_t i = states_.size() - static_cast<size_t>(max); i < states_.size(); i++) {
        const std::string line = states_[i]->str();
        std::fputs(line.c_str(), stderr);
        out("\x1b[K");
        if (i + 1 < states_.size()) {
            out("\n");
        }
    }
    pos_ = static_cast<int>(states_.size());
    out("\x1b[?25h");
    out("\x1b[?2026l");
    std::fflush(stderr);
}

void Progress::stop() {
    bool first = false;
    {
        std::lock_guard<std::mutex> lock(mu_);
        first   = !halted_;
        halted_ = true;
    }
    done_ = true;
    if (thread_.joinable()) {
        thread_.join();
    }
    if (!first) {
        return;
    }
    {
        std::lock_guard<std::mutex> lock(mu_);
        for (const auto & s : states_) {
            if (auto * sp = dynamic_cast<ProgSpinner *>(s.get())) {
                sp->stop();
            }
        }
        render_locked();
    }
    if (stderr_is_terminal()) {
        out("\n");
        std::fflush(stderr);
    }
}

void Progress::stop_and_clear() {
    int lines = 0;
    {
        std::lock_guard<std::mutex> lock(mu_);
        lines   = pos_;
        halted_ = true;
    }
    done_ = true;
    if (thread_.joinable()) {
        thread_.join();
    }
    if (!stderr_is_terminal()) {
        return;
    }
    out("\x1b[?25l");
    for (int i = 0; i < lines; i++) {
        if (i > 0) {
            out("\x1b[A");
        }
        out("\x1b[2K\x1b[1G");
    }
    out("\x1b[?25h");
    std::fflush(stderr);
    std::lock_guard<std::mutex> lock(mu_);
    pos_ = 0;
}

} // namespace llmash
